server/upsd.c

```diff
@@ -275,8 +277,169 @@ static void setuptcp(stype_t *server)
 	struct addrinfo		hints, *res, *ai;
 	int	v = 0, one = 1;
 
+	if (VALID_FD_SOCK(server->sock_fd)) {
+		/* Alredy bound, e.g. thanks to 'LISTEN *' handling and injection
+		 * into the list we loop over */
+		upsdebugx(6, "setuptcp: SKIP bind to %s port %s: entry already initialized",
+			server->addr, server->port);
+		return;
+	}
+
 	upsdebugx(3, "setuptcp: try to bind to %s port %s", server->addr, server->port);
 
+	/* Special handling for `LISTEN * <port>` directive with literal asterisk:
+	 * on systems with RFC-3493 (no relation!) support for "IPv4-mapped addresses"
+	 * it suffices to LISTEN on "::" (aka "::0" or "0:0:0:0:0:0:0:0") and also
+	 * get an IPv4 any-address listener. More so, they would conflict and
+	 * listening on one such socket precludes listening on the other. On other
+	 * systems (or with disabled mapping so IPv6 means IPv6 only) we need both.
+	 * So we jump through some hoops:
+	 * * Try to get IPv4 any-address, just to know if it is available right now;
+	 * * Free it and try to get IPv6 any-address, and try to get again that
+	 *   IPv4 any-address (IFF it was available before but is not available now -
+	 *   not a problem).
+	 * * Remember the entries used, to release later.
+	 * For more details see https://github.com/networkupstools/nut/issues/2012
+	 */
+	if (!strcmp(server->addr, "*")) {
+		stype_t	*serverAnyV4 = NULL, *serverAnyV6 = NULL;
+		int	canhaveAnyV4 = 0;
+		int	canhaveAnyV6 = 0;
+
+		/* For this use-case, we allow IPv6 to handle IPv4 if it can */
+		int	old_require_IPV6_V6ONLY = require_IPV6_V6ONLY;
+		require_IPV6_V6ONLY = (opt_af == AF_INET6);
+
+		/* Note: default opt_af==AF_UNSPEC so not constrained to only one protocol */
+		if (opt_af != AF_INET6) {
+			/* Not constrained to IPv6 */
+			upsdebugx(1, "%s: handling 'LISTEN * %s' with IPv4 any-address support",
+				__func__, server->port);
+			serverAnyV4 = xcalloc(1, sizeof(*serverAnyV4));
+			serverAnyV4->addr = xstrdup("0.0.0.0");
+			serverAnyV4->port = xstrdup(server->port);
+			serverAnyV4->sock_fd = ERROR_FD_SOCK;
+			serverAnyV4->next = NULL;
+		}
+
+		if (opt_af != AF_INET) {
+			/* Not constrained to IPv4 */
+			upsdebugx(1, "%s: handling 'LISTEN * %s' with IPv6 any-address support",
+				__func__, server->port);
+			serverAnyV6 = xcalloc(1, sizeof(*serverAnyV6));
+			serverAnyV6->addr = xstrdup("::0");
+			serverAnyV6->port = xstrdup(server->port);
+			serverAnyV6->sock_fd = ERROR_FD_SOCK;
+			serverAnyV6->next = NULL;
+		}
+
```

Comment:
This feels way too complicated.   I see two reasonable approaches for *:
1)
  - create a v6 socket and bind to ::0 (unless v6 is disabled at build time or command line)
  - create a v4 socket and bind to 0.0.0.0 (unless disabled).   If there's an error, log it.  People with v6only=0 getting an error is not a bug :-)
 2)
   as above, but setsockopt v6only=1 on the v6 socket

I now think 2 is the right answer.

With this, there is no need to have all this complexity.   Nut has a lot of lines of code, and less is more.  So we should try hard to minimize code, even at the expense of minor niceties for systems that aren't as they should be.  But I think with 2, it shouldn't cause anyone trouble.